APMSP-3550 DI: add re-entrancy guard for method probes on stdlib (#5560)
DataDog/dd-trace-rb: docs/DynamicInstrumentation.md

  (defined via `def self.method_name`, `class << self`, or `module_function`)
  cannot be instrumented with method probes. Line probes inside class
  methods still work since line probes are not method-bound.
- On Ruby 3.3 and later, method probes on `Kernel#lambda` are not supported

> due to Ruby language limitations.

Reviewer: Let's do it for all versions, it's a really bad idea to DI `Kernel#lambda`

Author: I will prohibit Kernel#lambda

Author: This is done in https://github.com/DataDog/dd-trace-rb/pull/5954

  it
- This threshold can be configured globally

## Probes on Standard Library Methods

Method probes can target standard library methods, including methods
that the Datadog tracer itself calls. Line probes can target any
line, including lines in standard library files.

The tracer guarantees that probes on such methods do not cause runaway
recursion or stack overflow. When the tracer calls a probed method
while processing another probe firing, the probe is suppressed for

> that internal call and no rate-limit token is consumed. Customer code

Reviewer: Is `rate-limit token` the correct terminology here?
It might be, I don't know, though it sounds like LLM stuff to me.
I would say:

```suggestion
that internal call and does not count towards the rate limit. Customer code
```

Author: You're right — "rate-limit token" leaks the token-bucket internals into a customer-facing doc. Applied the suggestion verbatim, and updated the matching comment in `instrumenter.rb:136` to use the same phrasing for consistency.

Fixed in:
- `docs/DynamicInstrumentation.md:341` — applied your suggestion verbatim (3eca65c8fc)
- `lib/datadog/di/instrumenter.rb:136` — updated the code comment to match (3eca65c8fc)

  it
- This threshold can be configured globally

## Probes on Standard Library Methods

Method probes can target standard library methods, including methods
that the Datadog tracer itself calls. Line probes can target any
line, including lines in standard library files.

The tracer guarantees that probes on such methods do not cause runaway
recursion or stack overflow. When the tracer calls a probed method
while processing another probe firing, the probe is suppressed for

> that internal call and does not count towards the rate limit. Customer

Author: Good catch — the docs guarantee was indeed broader than the guard's scope. `line_trace_point_callback` doesn't enter the guard, so a method probe on a stdlib method called while a line probe is processing its snapshot still fires.

Narrowed the docs rather than extending the guard. Line probes already self-disable via Ruby's TracePoint, so they don't recurse on themselves — the only remaining interaction is line probe + method probe on a tracer-used stdlib, which is now explicitly documented as not suppressed by the method-probe guard (but rate-limited normally). Also added a matching "Scope:" paragraph at the top of the wrapper in `instrumenter.rb` so source and docs agree.

Fixed in 0a3dae079:
- `docs/DynamicInstrumentation.md:365-377` — guarantee restated as "another method probe firing"; added a paragraph stating that line probes rely on TracePoint self-disabling and that a method probe on a stdlib called during line-probe snapshot building is not suppressed by the guard but is rate-limited
- `lib/datadog/di/instrumenter.rb:121-128` — added a "Scope:" paragraph stating the guard protects method-probe → method-probe re-entrancy only
